File: 1202. Smallest String With Swaps.py
class Solution:
    def smallestStringWithSwaps(self, s: str, pr: List[List[int]]) -> str:
        n, fa = len(s), [i for i in range(len(s))]
        
        def find(x: int):
            if fa[x]!=x: fa[x] = find(fa[x])
            return fa[x]
        
        for x, y in pr:
            fx, fy = find(x), find(y)
            if fx != fy:
                fa[fx] = fy
                
        cnt = dict()
        for i in range(n):
            f = find(i)
            if not f in cnt:
                cnt[f] = [0] * 26
            cnt[f][ord(s[i])-97] += 1
        
        p = collections.defaultdict(int)
        ans = ""
        for i in range(n):
            while cnt[fa[i]][p[fa[i]]]==0:
                p[fa[i]] += 1
            ans += chr(p[fa[i]] + 97)
            cnt[fa[i]][p[fa[i]]] -= 1
        
        return ans

        

File: 1202. Smallest String With Swaps.cpp
class Solution {
public:
    vector<int> fa;
    
    int find(int x) {
        return fa[x]==x ? x : fa[x] = find(fa[x]);
    }
    
    string smallestStringWithSwaps(string s, vector<vector<int>>& pr) {
        int n = s.size();
        fa.resize(n);
        iota(fa.begin(), fa.end(), 0);
        for (auto& p : pr) {
            int fx = find(p[0]), fy = find(p[1]);
            if (fx != fy) 
                fa[fx] = fy;
        }
        unordered_map<int, vector<int>> cnt;
        for (int i=0; i<n; ++i) {
            int f = find(i);
            if (!cnt.count(f))
                cnt[f] = vector<int>(26);
            ++cnt[f][s[i]-'a'];
        }
        unordered_map<int, int> p;
        for (int i=0; i<n; ++i) {
            int& pf = p[fa[i]];
            vector<int>& c = cnt[fa[i]];
            while (c[pf]==0) ++pf;
            s[i] = pf + 'a';
            --c[pf];
        }
        return s;
    }
};